### PR TITLE
Add USWDS alpha and gulp pipeline

### PR DESCRIPTION
**Pull request by thisisdano**
[Preview](https://federalist-proxy.app.cloud.gov/preview/gsa/10x/dw-add-uswds-alpha/)

### tldr
- Added USWDS Alpha Utilities
- Added gulp pipeline to process SASS
- Removed references to existing CSS

### What I was trying to do
This is  going to be a little longer than the usual PR description, as I'm trying to work a few things out, especially as relates to starting a "blank page" bare sandbox Jekyll USWDS site on Federalist. There actually aren't a ton of resources out there for doing this, and fewer since I'm test driving a new set of tools. 

There's a USWDS Jekyll template [here](https://github.com/18F/uswds-jekyll) and a simple SASS started project [here](https://github.com/18F/uswds-simple-sass) but neither was exactly what I'm looking for. The Jekyll template is currently too opinionated and is tied to an existing version of the design system that's not as applicable to this project. The simple sass project (thanks @toolness) is a good way to get started with a gulp pipeline, especially if you're trying to do some things with SASS that are more complicated than Jekyll's built-in SASS compiler can do on its own.

- build a large set of prototyping utilities
- build a smaller set of production utilities
- allow project-specific overrides to how these utilities are output
- allow custom subsetting
- output to nonstandard locations for inlining in the `head`
- develop something that is reusable on similar USWDS project

I want to be able to build with USWDS variables and utilities _without_ any of the existing USWDS CSS, and I want to be able to have the most performant production site possible (as audited by the performance section of Lighthouse).

### 1. Adding USWDS
How do you add USWDS to a project? For me, at least, there are a few challenges:
- What method do you use to actually add the files to a jekyll project?
- How do you set up your project files to reference and customize default USWDS?

#### What method do you use to actually add the files to your project?
I added the files with a very brute-force copy/paste from an existing `uswds/uswds` repo. Typically, though, I would add them via NPM with a `npm install uswds`, but since this is a pre-alpha version of USWDS, _and_ I was interested in using my install to make changes to the UWDS alpha, I couldn't do that. It might have made sense to add it as a git submodule, but I wasn't interested in going there. I just copy/pasted in to a `/vendor` folder and decided to treat it like `node_modules`.
```
project-root/
├── _vendor/
│   ├── uswds/
```

#### How do you set up your project files to reference and customize default USWDS?
Now, all the USWDS source files are in my project, and I need a way to access them. I don't want `_vendor` included in my Jekyll build (which is why it's prefixed with an underscore), so I need a way to access those source files and compile them into something usable, outside `_vendor`. I've done this with a gulpfile that can generate USWDS CSS from built-in USWDS defaults, or with project-specific settings files. I made a `LOCATIONS` section in my gulpfile to set common locations referenced by the build process
```
// USWDS source directory
const USWDS_SRC         = '_vendor/uswds';

// Project Sass source directory
const PROJECT_SASS_SRC  = '_sass';

// Asset destination
const ASSETS_DEST       = 'assets/uswds';

// CSS destination
const CSS_DEST          = 'assets/css';

// Build destination
const BUILD_DEST        = '_site';

// Include destination
const INC_DEST          = '_includes';
```

I'll reference these locations for the build tasks, like
```
gulp.task('build-prototyping-utilities', function (done) {
  return gulp.src([
      `${USWDS_SRC}/stylesheets/uswds-fonts.scss`,
      `${USWDS_SRC}/stylesheets/uswds-prototyping-utilities.scss`
    ])
    ...
    .pipe(gulp.dest(`${CSS_DEST}`));
});
```

The goal was to build a set-it-and-forget-it build where I could set some variables at the top and the final CSS would just appear where Jekyll could find and use it.

This could compile the USWDS source from `_vendor` but it still didn't have a way to manually customize USWDS project settings without modifying the USWDS source. Many USWDS settings are suffixed with `!default` meaning that the value is set only if a value for that variable has not previously been set. So, we need a way to load project variables and settings _before_ the default values in the build. As always, there are some quirks.

USWDS source files need to `import` in a specific order so the necessary variables are available. In the following example, the system calculates the rem value of a unit in the global spacing palette, and then a developer can assign that value to a project setting:

1. `$grid-base` is set to `8px !default`
2. `$font-size-base` is set to `10px !default`
3. The `spacing(n)` function is defined as `n * $grid-base / $font-size-base`rem [**not modifiable**]
4. The unit `$u-spacing-2` unit is set to `spacing(2)` [**not modifiable**]
5. The setting `$s-spacing-md` is set to `$u-spacing-2 !default`

So you can imagine an import order like
```
@import base-values
@import functions
@import unit-definitions
@import settings
```

We can't just add a settings import at the top, like 
```
@import project-settings
@import base-values
@import functions
@import unit-definitions
@import settings
```
This is because the global spacing palette (like `$u-spacing-2`) hasn't been defined yet. It won't be defined until `@import unit-definitions`. Thus we can't assign `$s-spacing-md` to `$u-spacing-2`! And we can't add a single `@import project-settings` after importing the unit definitions either, since the project settings need to be able to modify the base values so USWDS can play well with sites that use a different base font size than `10px`.

So, we need _two_ separate project settings files. I've named them `_uswds-project-init.scss` and `uswds-project-settings`. I've included the in a new folder inside the `uswds` root called `project`.
```
@import uswds-project-init // to set the base values
@import base-values
@import functions
@import unit-definitions
@import uswds-project-settings // to set the project settings
@import settings
```

To use these settings files, a new project developer would need to copy them from the `uswds/project` folder into their project sass source directory (or `PROJECT_SASS_SRC` in the gulpfile. (They'd also need to copy the `.gulpfile` itself from the uswds root into their own project root.) 

Still it's not enough! The sass compiler needs to know to use the project's settings files and not those included in the USWDS distribution. We can set the proper order with `includePaths`:
```
.pipe(sass({
  includePaths: [
    `${PROJECT_SASS_SRC}`,
    `${USWDS_SRC}/stylesheets/project`,
  ]
}).on('error', sass.logError))
```
Now, our overrides will load (if they exist) and the default settings will only load if our project settings do not. 

### 2. Typefaces
Custom typography makes USWDS more complicated than other systems that don't offer typeface choice. We want a resilient type solution that to assures we support the best type for the job, that any solution can work well in an environment when custom typography doesn't load, and that projects that use custom type can easily be integrated back into the system.

This is still a work in progress, but USWDS Alpha supports font maps, sass maps that collect typeface-specific information for programmatic output. The following example shows the kind of information included in a font map:

```
$font-united-sans : (
  name: "United Sans Web",
  dir: "united-sans",
  stack: $helvetica,
  system-font: false,
  500px-cap-height: 362px,
  variable-font: false,
  roman: (
    100: "UnitedSans-Thin",
    200: "UnitedSans-ExtraLight",
    300: "UnitedSans-Light",
    400: "UnitedSans-Regular",
    500: "UnitedSans-Medium",
    600: "UnitedSans-SemiBold",
    700: "UnitedSans-Bold",
    800: "UnitedSans-ExtraBold",
    900: "UnitedSans-Heavy",
  ),
  italic: (
    100: "UnitedSans-ThinItalic",
    200: "UnitedSans-ExtraLightItalic",
    300: "UnitedSans-LightItalic",
    400: "UnitedSans-Italic",
    500: "UnitedSans-MediumItalic",
    600: "UnitedSans-SemiBoldItalic",
    700: "UnitedSans-BoldItalic",
    800: "UnitedSans-ExtraBoldItalic",
    900: "UnitedSans-HeavyItalic",
  ),
);
```

This allows us to do a couple different things. We can use the `500px-cap-height` value to normalize a typeface to a normal value — metrically smaller faces will display larger and metrically large faces will display smaller. Faces would be normalized around a baseline size common to the vast majority of users and operating systems, like Helvetica/Arial. Additionally, having all this information together allows the sass compiler to combine multiple project settings into the correct `@font-face` declarations. Given the following settings:
```
$s-font-sans:   $font-united-sans;
$s-font-serif:  $font-georgia;
$s-font-mono:   $font-roboto-mono;

$s-fw-xlight:             200;
$s-fw-light:              300;
$s-fw-normal:             400;
$s-fw-bold:               600;
$s-fw-xbold:              700;
```
The builder will only declare @font-faces for United Sans and Roboto Mono in the declared weights, if that weight is available in the face. These `@font-face` declarations are bundled into a separate SASS file `uswds-fonts.scss` that is, by default, concatenated into the main app css.

However, that `@font-face` declaration file can be maintained as a separate file, whose loading can be deferred as a performance optimization, as we will see.

To better support custom type, I've included Merriweather, Roboto Mono, Source Sans Pro, and United Sans in all the common webfont formats in separate folders within `uswds/fonts`. Currently all the faces are copied over as project assets, but only the requested faces are referenced as `@font-face`s.

### 3. Optimizing USWDS
I'm currently focussed on optimizing a utility-first workflow based on the Alpha utilities.

Like any CSS, utilities benefit from subsetting. Even a complex site with sophisticated functionality would use significantly fewer utility classes than we provide in our Prototyping Utilities set. I hypothesize that a site built out with either Production Utilities or Prototyping Utilities could be subset to an unzipped size of under 5 KB. 

At this weight, it would be reasonable to include the full CSS inline in the `head` of the site html, removing necessity of render-blocking stylesheets. However, this can cause a bit of an issue when referencing fonts or images from the CSS — depending on the place in the directory tree where the final page lives, the relative path to the fonts and images will change. There's probably some smart way around this! However, I've decided to eschew that smart solution for a different one — perhaps one that's smart in a different way. 

I've split the `@font-face` declarations CSS into a separate file — `uswds-fonts.scss`. If you set the `CONCAT_FONTS` constant to `false` in the gulpfile, this separate file will not be concatenated into final app css. The final CSS builds (`gulp uswds-app`) will result in a minified concatenation of any of the modules set in `USWDS_APP`:
```
const USWDS_APP = [
  "build-production-utilities",
  "build-prototyping-utilities",
//  "build-uswds"
];
```

Furthermore, `gulp uswds-opt` will subset this concatenation against the html in `_site` and place the final optimized css `uswds-app-opt.min.css` in `/includes` so it can be included  inline as `<style>{% include uswds-app-opt.min.css %}</style>` in the `head`.

This technique results in all necessary CSS _except the @font-face declarations_ loading in the head. 
Then we can use the [LoadCSS technique](https://github.com/filamentgroup/loadCSS) from the Filament Group to use inline JavaScript to load the @font-face declarations asynchronously! This results in a FOUT, but since our typefaces progressively enhanced to be normalized against the stack fallbacks, the impact of the FOUT is pretty minimal — _and_ if either the fonts don't load or the JavaScript is blocked, the site should be perfectly handsome and functional even without custom typography.

This technique is a reliable way to get a **>90** lighthouse performance audit score — which is kinda tough to get!

#### A note about CSS compile times
With all this shenanigans going on behind the scenes, there's going to be a toll on CSS compile times. The complete optimization process takes about 25 seconds, which is way too long for a typical day-to-day workflow. Fortunately, all of this optimization is only necessary before a final push to production. Local development, running on the full, unoptimized CSS needn't build CSS at all. Using utility classes means all the heavy lifting takes place in the templates and for a small site, incremental Jekyll builds with live reload (`jekyll serve --livereload --incremental`) are almost instantaneous. 

For standard development work, set `optimize_css: false` in `_config.yml`. This points the css to
```
<link rel="stylesheet" href="{{ "/assets/css/uswds-fonts.min.css" | relative_url }}">
<link rel="stylesheet" href="{{ "/assets/css/uswds-app-all.min.css" | relative_url }}">
```
And there's no inlining or javascript font-face loading. It's fast.

Run `gulp uswds-opt` and set `optimize_css: true` before testing on Federalist.

### What's next? (Not necessarily in this PR)
- ~~Adding a place for custom CSS~~  **DONE!**
- Building custom CSS with settings and unit vars, OR building them from utility @includes (which I don't really want to do).
- Building and referencing components
- Implementing font-awesome and icons
- Testing settings utilities
- Building default and watch gulp tasks
- Converting gulp tasks into npm scripts
- ~~Determining if any of this can be done automatically in Federalist~~ **DONE!**
- More work around typefaces
- More work around settings and customization

### CODE REVIEW

**Comment by jeremyzilar** [comment]
❤️ This is a good read

**Comment by thisisdano** [comment]
screen shot 2018-02-27 at 3 50 01 pm

Lighthouse results. Not too shabby. **Best Practices** should improve once we get on a real URL over HTTP2.
